### DISCUSSION TITLE
How do you provide public stubs your packages?

### DISCUSSION THREAD

**Discussion by jvrana** [microsoft/pyright#3923] [Q&A]
I have a number of interconnect packages I am working on. On import of my packages (call it **Pkg**) into my other project (call it **MainPkg**) I am getting "error -Type of OBJECT is unknown" for any objects imported from my library. I know I can quell these errors by generating stubs in my **MainPkg** using `pyright --createstub Pkg` which will generate the stubs in the **MainPkg** `typings/pkg/**/*.pyi`

_Is this before expected?_ My **Pkg** library has all type defined so I was a bit surprised when on import, I still needed to generate stubs for the Pkg. Is so, that's fine I can generate stubs.
 
_Where do I put the stubs for **Pkg**?_ Is there a way to provide the stubs in **Pkg** so I do not have to run `pyright --createstubs` for every single project uses **Pkg** as a dependency? I've created stubs in various locations Pkg/stubs, Pkg/pkg-stubs Pkg/typings/pkg/ and included them in the build before installing (and confirming they are included). Pyright seems unable to find these stubs on these build wheels.

More broadly, I guess I'm asking how Pyright looks for stubs for third-party applications? I know its looking in the main package under typings/**/*.pyi and some typeshed for standard python libraries. Where else does it look when importing a package? I'm having difficulty finding guidance for how to package up stubs. 

Thanks,
Justin

**Comment by erictraut** [comment]
Am I correct in understanding that *Pkg* in your example is a library that is intended to be installed and consumed by other projects? 

You mentioned that *Pkg* already has all types defined. Does it include a "py.typed" marker file as specified in [PEP 561](https://peps.python.org/pep-0561/#packaging-type-information)? I'm guessing that if you simply add this file, it will work as you're hoping. You might find [this documentation](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) useful.

If you don't include a "py.typed" marker, pyright assumes by default that any symbols imported from that library should be of "Unknown" type. It also offers a mode that can be enabled via the `--lib` command-line flag or the `useLibraryCodeForTypes` configuration setting. When this mode is enabled, pyright will use declared inlined types (and infer missing types as necessary) from a library even if it's not marked "py.typed".
